fix for issue #2430

gfelber opened the pull request:
<!-- Please make sure to read the testing and linting instructions at https://github.com/pwndbg/pwndbg/blob/dev/DEVELOPING.md before creating a PR -->

fixes #2430, that compares base file names instead of absolute paths for `pwndbg.gdblib.proc.binary_vmmap`, if the target is run on remote.

Aplet123 commented:
This seems janky to me, I think it's preferable to pull the name from `pwndbg.aglib.proc.exe` instead and do an exact comparison.

gfelber commented:
is the new solution preferred? 
Also, sry for the commit spam i don't have a linter in my nvim setup

Aplet123 commented:
Have you tested the new code? I think it might not work, as it doesn't remove the `target:` prefix from the exe name, and still gets the basename of the objfile.

gfelber commented:
~~it worked on my setup (explained in issue)~~

gfelber commented:
nvmd it didn't work

gfelber commented:
pwndbg.aglib.proc.exe still returns the absolute path (and not `target:` prefix),
do you want this comparison
```python
                if os.path.basename(p.objfile) == os.path.basename(pwndbg.aglib.proc.exe)
```

gfelber commented:
these are my gdb settings
```
 ['/usr/bin/gdb', '-ex', 'set sysroot /PATH/TO/SYSROOT/./.vagd/sysroot/', '-ex', 'file -readnow ./binary', '-q', '-x', '/tmp/pwnlib-gdbscript-f6on
dmql.gdb']
```
note that I set file to my local path (so i don't need to redundantly download it from the remote.
Not sure if this creates a misunderstanding, but might be interesting information

Aplet123 commented:
I think `p.objfile == pwndbg.aglib.proc.exe.removeprefix("target:")` should work.

gfelber commented:
no sadly not, pwndbg.aglib.proc.exe returns the absolute path on my local system, not the remote

Aplet123 commented:
Can you send the result of `info proc mappings` and `pi pwndbg.aglib.proc.exe`?

gfelber commented:
`info proc mappings`:
```
Mapped address spaces:

          Start Addr           End Addr       Size     Offset  Perms  objfile
            0x400000           0x401000     0x1000        0x0  r--p   /REMOTE/PATH/binary
           0x8000000          0x8001000     0x1000     0x1000  r-xp   /REMOTE/PATH/binary
           0x8001000          0x8002000     0x1000     0x2000  rw-p   /REMOTE/PATH/binary
          0xffff0000         0xfffff000     0xf000        0x0  rw-p   
      0x7ffff7ff9000     0x7ffff7ffd000     0x4000        0x0  r--p   [vvar]
      0x7ffff7ffd000     0x7ffff7fff000     0x2000        0x0  r-xp   [vdso]
      0x7ffffffde000     0x7ffffffff000    0x21000        0x0  rw-p   [stack]
```
`pi pwndbg.aglib.proc.exe`
```
/LOCAL/PATH/binary
```

gfelber commented:
> these are my gdb settings
> 
> ```
>  ['/usr/bin/gdb', '-ex', 'set sysroot /PATH/TO/SYSROOT/./.vagd/sysroot/', '-ex', 'file -readnow ./binary', '-q', '-x', '/tmp/pwnlib-gdbscript-f6on
> dmql.gdb']
> ```
> 
> note that I set file to my local path (so i don't need to redundantly download it from the remote. Not sure if this creates a misunderstanding, but might be interesting information

this might be the reason `pi pwndbg.aglib.proc.exe` returns the local path

gfelber commented:
yeah if you specify the file locally using args `gdb ./binary` or the file command in gdb `file ./binary` pwndbg.aglib.proc.exe get's overwritten with the local path

gfelber committed:
gdblib & aglib proc now returns executed filename

+ executed filename received through "info proc exe"
+ previous the local filename was returned, which could be different from remote
+ this also fixes #2430
+ fixed softlock if `set integration-provider binja`

gfelber commented:
also fixed a soft lock issue if `set integration-provider binja` is executed before the binary is run

disconnect3d commented:
@Aplet123 Wdyt?

disconnect3d commented on pwndbg/aglib/proc.py:
```diff
@@ -75,18 +75,15 @@ def alive(self) -> bool:
-        See this by executing those in two terminals:
-        1. gdbserver 127.0.0.1:1234 /bin/ls
-        2. gdb -ex "target remote :1234" -ex "pi pwndbg.gdblib.proc.exe"
+        On remote targets, this path most definitly won't exist locally.
```
```suggestion
        On remote targets, this path may not exist locally.
```

disconnect3d committed:
Update pwndbg/aglib/proc.py

disconnect3d commented on pwndbg/dbg/gdb.py:
```diff
@@ -738,7 +738,11 @@ def module_section_locations(self) -> List[Tuple[int, int, str, str]]:
+        if not self.alive():
+            return gdb.current_progspace().filename
+
+        exe = gdb.execute("info proc exe", to_string=True)
```
Fwiw, just for information, it seems GDB fetches this via `/proc/$pid/exe`:

https://github.com/bminor/binutils-gdb/blob/gdb-15.1-release/gdb/infcmd.c#L3024-L3028

https://github.com/bminor/binutils-gdb/blob/gdb-15.1-release/gdb/infcmd.c#L2960-L2972


https://github.com/bminor/binutils-gdb/blob/gdb-15.1-release/gdb/linux-tdep.c#L810-L821

https://github.com/bminor/binutils-gdb/blob/gdb-15.1-release/gdb/linux-tdep.c#L882-L891

gfelber replied:
make sense, but creating a custom implementation that simulates this behavior (for remote and local) would be worse imo (way to complex for what it's worth)
therefore i would believe that `info proc exe`, is still the preferred solution

disconnect3d commented on pwndbg/gdblib/proc.py:
```diff
@@ -107,20 +107,34 @@ def stopped_with_signal(self) -> bool:
-        See this by executing those in two terminals:
-        1. gdbserver 127.0.0.1:1234 /bin/ls
-        2. gdb -ex "target remote :1234" -ex "pi pwndbg.gdblib.proc.exe"
+        On remote targets, this path most definitly won't exist locally.
```
```suggestion
        On remote targets, this path may not exist locally.
```

disconnect3d committed:
Update pwndbg/gdblib/proc.py

disconnect3d commented on pwndbg/gdblib/proc.py:
```diff
@@ -107,20 +107,34 @@ def stopped_with_signal(self) -> bool:
+        If you need the locally referenced file use:
+            `gdb.current_process().filename`
+
+        info proc exe is the only command to actually get the executed file.
```
```suggestion
        info proc exe is the only command to actually get the executed file path.
```

disconnect3d committed:
Update pwndbg/gdblib/proc.py

patryk4815 commented:
should we write tests to check this behavior? (`exe`)

disconnect3d commented:
Ideally, yes pls

Aplet123 commented:
> @Aplet123 Wdyt?

LGTM, although I wonder what the observed deadlock is with enabling binja integration before starting the binary is, as I can't seem to reproduce it.

gfelber commented:
weird can't seem to reproduce the binja integration softlock bug again either, maybe it was some issue that occurred while I messed with the code base.

I just noticed that the pwndbg event handler trigger an update that resulted in calling `gdblib.proc.get_base_adddress` while the process wasn't running, causing an error and preventing gdb from any state changes (because they triggered the event handler again, which caused another crash)

gfelber commented:
I think i recreate the bug (or found another one), you get the same issue if you setup the integration and then stop it inside binary ninja, it doesn't seem to actually softlock gdb, but it breaks the context view (and generally speaking all pwndbg commands), you can't even disable it afterwards with `set integration-provider none`

gfelber commented:
yeah this is not the issue i had yesterday, but an issue nonetheless

gfelber committed:
fixed binja integration issues

+ ensure current process is alive, for needed events
+ handle ConnectionErrors correctly

gfelber committed:
fixed binja integration issues

+ ensure current process is alive, for needed events
+ handle ConnectionErrors correctly

gfelber commented:
ok implemented fixes for the "new" issue that should at least prevent the pwndbg soft lock issue. 
Also kept the `gdblib.proc.alive` requirement for events that actually need it to be alive, not sure if this is needed, but it makes sense to check it and safely return, to prevent potential pwndbg soft locks.

disconnect3d commented on pwndbg/integration/binja.py:
```diff
@@ -153,13 +153,24 @@ def init_bn_rpc_client() -> None:
+
         @functools.wraps(func)
         def wrapper(*args: P.args, **kwargs: P.kwargs) -> T | K:
+            global _bn
```
Does this really require global x3? I think only this one is needed?

gfelber replied:
i tried but it didn't work without it, just python stuff i guess
